packages/core/src/sync-realtime/service.ts

```diff
@@ -548,106 +516,119 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
     return { blocks: newBlocks, logs: matchedLogs, reorg: false };
   };
 
-  private syncBatch = async (
-    newBlock: RealtimeBlock,
-  ): Promise<
-    | {
-        blocks: RealtimeBlock[];
-        logs: RealtimeLog[];
-        reorg: boolean;
-      }
-    | { reorg: true }
-  > => {
-    const latestLocalBlock = this.getLatestLocalBlock();
-    const latestLocalBlockNumber = latestLocalBlock.number;
+  /**
+   * Find common ancestor block and evict local blocks and logs that has been reorged out of the chain.
+   */
+  reconcileReorg = async () => {
+    const hasDeepReorg = await this.validateFinalizedBlock();
+    if (hasDeepReorg) return;
+
+    // Find common ancestor block by requesting remote block until you find one that fits in the local chain.
+    while (this.blocks.length > 0) {
+      const localBlock = this.blocks.pop();
+      const remoteBlock = await this._eth_getBlockByNumber(localBlock!.number);
+      const parent =
+        this.blocks.length === 0
+          ? this.finalizedBlock
+          : this.blocks[this.blocks.length - 1];
+
+      if (parent.hash === remoteBlock.parentHash) {
+        this.common.logger.trace({
+          service: "realtime",
+          msg: `Found common ancestor block ${parent.number} (network=${this.network.name})`,
+        });
 
-    const newBlockNumber = hexToNumber(newBlock.number);
+        this.logs = this.logs.filter((log) => log.blockNumber <= parent.number);
 
-    if (latestLocalBlockNumber >= newBlockNumber) {
-      return { reorg: true };
-    }
+        await this.syncStore.deleteRealtimeData({
+          chainId: this.network.chainId,
+          fromBlock: BigInt(parent.number),
+        });
 
-    this.common.logger.trace({
-      service: "realtime",
-      msg: `Requesting logs from ${
-        latestLocalBlockNumber + 1
-      } to ${newBlockNumber} (network=${this.network.name})`,
-    });
+        this.emit("reorg", {
+          blockTimestamp: parent.timestamp,
+          chainId: this.network.chainId,
+          blockNumber: parent.number,
+        });
 
-    // Get logs
-    const _logs = await this._eth_getLogs({
-      fromBlock: numberToHex(latestLocalBlockNumber + 1),
-      toBlock: newBlock.number,
-    });
-    const logs = sortLogs(_logs);
+        this.common.logger.warn({
+          service: "realtime",
+          msg: `Detected reorg with common ancestor ${parent.number} (network=${this.network.name})`,
+        });
 
-    const matchedLogs = await this.getMatchedLogs(
-      logs,
-      BigInt(newBlockNumber),
-      true,
+        return;
+      }
+    }
+
+    throw new Error(
```

Comment:
Why is this an invariant violation?

Reply:
The code should never get to this point because:

- we have a complete local chain
- we first check for deep reorg
- if not deep reorg we know the local chain is at least partly in tact
- we find the common ancestor by stepping through the chain backwards, re-requesting every block and seeing if the remote block fits into the local chain for that block number
- getting to this line means the local chain is in tact but we weren't able to find a common ancestor in the local chain, which is an invariant violation